File: CSharp/PinetreeChat.DataAccess/Repositories/UserRepository.cs
using PinetreeChat.Interfaces.Repositories;
using System.Collections.Generic;
using PinetreeChat.Entities;
using System.Linq;

namespace PinetreeChat.DataAccess.Repositories
{
    public class UserRepository : IUserRepository
    {
        private static List<User> _users = new List<User>();

        public void AddUser(User user)
        {
            _users.Add(user);
        }

        public ICollection<User> GetUsers()
        {
            return _users.ToList();
        }

        public void SetLoggedIn(User user)
        {
            user.IsLoggedIn = true;
        }

        public void SetLoggedOut(User user)
        {
            user.IsLoggedIn = false;
        }
    }
}

File: FSharp/PinetreeChat.DataAccess/UserRepository.fs
namespace PinetreeChat.DataAccess

module UserRepository =
    open PinetreeChat.Entities

    let mutable private userList = List.empty<User>
    
    let private updateUserInList updateFunction username =
        let mappingFunction user = 
            if (user.Username = username)
            then updateFunction user
            else user           
            
        userList <- List.map mappingFunction userList
        List.find (fun u -> u.Username = username) userList

    let getUsers () =
        userList 

    let getUser username =
        try List.find (fun u -> u.Username = username) userList |> Some
        with _ -> None

    let addUser user =
        userList <- user :: userList
        user

    let setLoggedIn =
        updateUserInList (fun u -> { u with IsLoggedIn = true })

    let setLoggedOut =
        updateUserInList (fun u -> { u with IsLoggedIn = false })
